WordPress/gutenberg
# Refactored `wp-env` Config Parsing

## What?
This pull request refactors the existing `.wp-env.json` config file parsing in order to more cleanly support root properties. It aimed to increase composition and testability while maintaining full compatibility.

## Why?
While working on https://github.com/WordPress/gutenberg/pull/49996 we decided that it would be best if the `postInstall` script existed at the configuration root. When I looked into implementing that, however, it proved to be troublesome due to the way config parsing was designed. It was explicitly created to cascade the config file almost immediately, and as a result, it was not easy to introduce root-level options and validation in a way that would not cascade.



## How?
We delay the cascade of options from the root to the environments until after validation and parsing have completed. This lets us interact with the config object while the root and environments are both present but separate. As a result we will be able to support root-level configuration options without any additional changes.

I was worried about breaking compatibility, so, this pull request also features an exhaustive test suite update. All new functionality is more or less completely covered for both happy path and failure states. This was a great exercise and caught a number of bugs in the process.

## Testing Instructions

1. Run `npm run env -- start --update` and make sure that both the development and tests environments work.
2. Run `npm run test:unit packages/env` and make sure that they pass.
3. Create a `.wp-env.override.json` file and make sure that it overrides as expected. Config and mappings should be merged while plugins and themes are overridden.
